k2/csrc/array.h

```diff
@@ -410,10 +417,46 @@ class Array2 {
     that
      it will have a different memory layout than the input.
   */
-  Array2<T> To(ContextPtr ctx) {
-    // TODO
-    Array2<T> array;
-    return array;
+  Array2<T> To(ContextPtr ctx) const {
+    if (dim0_ * dim1_ == 0) {
+      if (!region_) return *this;
+
+      if (ctx->IsCompatible(*Context())) return *this;
+
+      Array2 ans(ctx, dim0_, dim1_);
+      return ans;
+    }
+
+    if (ctx->IsCompatible(*Context()) && elem_stride0_ == dim1_) return *this;
+
+    // the current array is either non-contiguous or not compatible with the ctx
+
+    Array2<T> ans(ctx, dim0_, dim1_);
+
+    if (elem_stride0_ == dim1_) {
+      // the current array is contiguous, use memcpy
+      auto kind = GetMemoryCopyKind(*Context(), *ctx);
+      auto *dst = ans.Data();
+      const auto *src = Data();
+      MemoryCopy(static_cast<void *>(dst), static_cast<const void *>(src),
+                 dim0_ * dim1_ * ElementSize(), kind);
+      return ans;
+    } else {
+      // do rowwise copy
```

Comment:
this will be super slow if you do sequential CUDA copies.  Instead, I'd return this->ToContiguous().To(ctx); solve it by recursion.

Reply:
Could you give some hint to implement `ToContiguous` without sequential copy?

Reply:
I guess something like using `kernel`, see my implementation in `Flatten`

Reply:
@qindazhu 
thanks. fixed.